src/flp/szk.rs

```diff
@@ -57,29 +60,144 @@ pub enum SzkError {
+                leader_blind_and_helper_joint_rand_part_opt: is_random.then_some((
+                    Seed::<SEED_SIZE>::decode(bytes)?,
+                    Seed::<SEED_SIZE>::decode(bytes)?,
+                )),
+            }),
+            1 => Ok(SzkProofShare::Helper {
+                proof_share_seed_and_blind: Seed::<SEED_SIZE>::decode(bytes)?,
+                leader_joint_rand_part_opt: is_random.then_some(Seed::<SEED_SIZE>::decode(bytes)?),
+            }),
+            _ => Err(CodecError::UnexpectedValue),
+        }
+    }
+}
+
+impl<F: Encode, const SEED_SIZE: usize> Encode for SzkProofShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match self {
+            SzkProofShare::Leader {
+                uncompressed_proof_share,
+                leader_blind_and_helper_joint_rand_part_opt,
+            } => (
+                uncompressed_proof_share.encode(bytes)?,
+                if let Some((blind, helper_joint_rand_part)) =
+                    leader_blind_and_helper_joint_rand_part_opt
+                {
+                    blind.encode(bytes)?;
+                    helper_joint_rand_part.encode(bytes)?;
+                },
+            ),
+            SzkProofShare::Helper {
+                proof_share_seed_and_blind,
+                leader_joint_rand_part_opt,
+            } => (
+                proof_share_seed_and_blind.encode(bytes)?,
+                if let Some(leader_joint_rand_part) = leader_joint_rand_part_opt {
+                    leader_joint_rand_part.encode(bytes)?;
+                },
+            ),
+        };
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        match self {
+            SzkProofShare::Leader {
+                uncompressed_proof_share,
+                leader_blind_and_helper_joint_rand_part_opt,
+            } => Some(
+                uncompressed_proof_share.encoded_len()?
+                    + if let Some((blind, helper_joint_rand_part)) =
+                        leader_blind_and_helper_joint_rand_part_opt
+                    {
+                        blind.encoded_len()? + helper_joint_rand_part.encoded_len()?
+                    } else {
+                        0
+                    },
+            ),
+            SzkProofShare::Helper {
+                proof_share_seed_and_blind,
+                leader_joint_rand_part_opt,
+            } => Some(
+                proof_share_seed_and_blind.encoded_len()?
+                    + if let Some(leader_joint_rand_part) = leader_joint_rand_part_opt {
+                        leader_joint_rand_part.encoded_len()?
+                    } else {
+                        0
+                    },
+            ),
+        }
+    }
+}
+
+impl<F: PartialEq, const SEED_SIZE: usize> PartialEq for SzkProofShare<F, SEED_SIZE> {
+    fn eq(&self, other: &SzkProofShare<F, SEED_SIZE>) -> bool {
```

Comment:
Why do we need `PartialEq` for `SzkProofShare`? Can it be derived instead?
```rust
#[derive(PartialEq)]
pub SzkProofShare<F, const SEED_SIZE: usize> {
   ...
}
```

Reply:
We need PartialEq for MasticInputShare, which contains a SZKProofShare. But yes, deriving it does seem sufficient.